### PR TITLE
Generating the analysis files 

### PR DESCRIPTION
**Pull request by IsaackMwenda**

### CODE REVIEW

File: project_test/analysis_file.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+from project_test.lib.test_schemes import CodeSchemes
+
+class ConsentUtils(object):
+    @staticmethod
+    def td_has_stop_code(td, coding_plans):
+        """
+        Returns whether any of the values for the given keys are Codes.STOP in the given TracedData object.
+        
+        Arguments:
+            td {TracedData} -- TracedData object to search for stop codes.
```
Comment (as2388) [inline project_test/analysis_file.py]: Why did you change the format of the docstrings?

File: project_test/analysis_file.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            data {iterable of TracedData} -- TracedData objects to set to stopped if consent has been withdrawn.
+        
+        Keyword Arguments:
+            withdrawn_key {str} -- Key in each TracedData object which indicates whether consent has been withdrawn. (default: {"consent_withdrawn"})
+        """
+
+        for td in data:
+            if td.get(withdrawn_key) == Codes.TRUE:
+                stop_dict = {key: Codes.STOP for key in td.keys() if key != withdrawn_key}
+                td.append_data(stop_dict, Metadata(user, Metadata.get_call_location(), time.time()))
```
Comment (as2388) [inline project_test/analysis_file.py:75]:
```suggestion
                td.append_data(stop_dict, Metadata(user, Metadata.get_call_location(), time.time()))       
                
                
```

File: project_test/analysis_file.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        
+        # Convert survey codes to their string values
+        for td in data:
+            td.append_data(
+                {plan.analysis_file_key: plan.code_scheme.get_code_with_id(td[plan.coded_field]["CodeID"]).string_value
+                for plan in PipelineConfiguration.SURVEY_CODING_PLANS},
+                Metadata(user, Metadata.get_call_location(), time.time())
+            )
+
+
```
Comment (as2388) [inline project_test/analysis_file.py]:
```suggestion
```

File: test_pipeline.py [hunk 4]
```diff
@@ -125,15 +126,18 @@
...
     data = AutoCodeShowMessages.auto_code_show_messages(user, data, icr_output_dir, coded_dir_path)
 
-    print("Exporting production CSV...")
-    data = ProductionFile.generate(data, production_csv_output_path)
-    
     print("Auto coding surveys...")
     data =AutoCodeSurveys.auto_code_surveys(user, data, phone_number_uuid_table_path, coded_dir_path)
 
+    print("Exporting production CSV...")
+    data = ProductionFile.generate(data, production_csv_output_path)
```
Comment (as2388) [inline test_pipeline.py]: Why did this need to move to be after the survey auto-coding?

File: project_test/analysis_file.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            for code in plan.code_scheme.codes:
+                show_matrix_keys.add(f"{plan.analysis_file_key}{code.string_value}")
+
+            AnalysisKeys.set_matrix_keys(
+                user, data, show_matrix_keys, plan.code_scheme, plan.coded_field, plan.analysis_file_key)
+
+            matrix_keys.extend(show_matrix_keys)
+        
+        matrix_keys.sort()
+
```
Comment (as2388) [inline project_test/analysis_file.py]:
```suggestion
```

File: project_test/analysis_file.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                plan.binary_code_scheme.get_code_with_id(td[plan.binary_coded_field]["CodeID"]).string_value
+                for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS if plan.binary_code_scheme is not None},
+                Metadata(user, Metadata.get_call_location(), time.time())
+            )
+        
+        # Translate the Test Show Question to matrix values
+        matrix_keys = []
+
+        for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS:
+            show_matrix_keys = set()
```
Comment (as2388) [inline project_test/analysis_file.py]: (1/2) I recently changed this to a list in ADSS. This means you get the codes in scheme order rather than alphabetical order...

File: project_test/analysis_file.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            for code in plan.code_scheme.codes:
+                show_matrix_keys.add(f"{plan.analysis_file_key}{code.string_value}")
+
+            AnalysisKeys.set_matrix_keys(
+                user, data, show_matrix_keys, plan.code_scheme, plan.coded_field, plan.analysis_file_key)
+
+            matrix_keys.extend(show_matrix_keys)
+        
+        matrix_keys.sort()
```
Comment (as2388) [inline project_test/analysis_file.py]: (2/2) ... and therefore you would no longer need this line

File: project_test/analysis_file.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        binary_keys = [plan.binary_analysis_file_key
+                        for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS
+                        if plan.binary_analysis_file_key is not None]
+        
+        equal_keys = ["uid"]
+        equal_keys.extend(survey_keys)
+        concat_keys = [plan.raw_field for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS]
+        bool_keys = [
+            consent_withdrawn_key,
+
```
Comment (as2388) [inline project_test/analysis_file.py]:
```suggestion
```

File: project_test/analysis_file.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+                    td.append_data({f"{plan.analysis_file_key}{Codes.NOT_CODED}": Codes.MATRIX_1},
+                                    Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+            
+            # Process consent
+            ConsentUtils.set_stopped(user, data, consent_withdrawn_key)
+            ConsentUtils.set_stopped(user, folded_data, consent_withdrawn_key)
+
+            # Output to CSV with one message per row
+            with open(csv_by_message_output_path, "w") as f:
```
Comment (as2388) [inline project_test/analysis_file.py]: Is this at the correct level of nesting? I think you're re-writing this file many, many times

**Review by as2388 — CHANGES_REQUESTED** [review]

File: project_test/analysis_file.py (hunk 1 above)
Comment (IsaackMwenda) [reply project_test/analysis_file.py]: I found it easier to write and interpret. Happy to revert to the former.

File: project_test/analysis_file.py (hunk 9 above)
Comment (IsaackMwenda) [reply project_test/analysis_file.py]: Thank you for noticing this.

**Comment by IsaackMwenda** [comment]
@as2388  PTAL! whats your advise should I keep the docstrings format as is or revert to the former.I do find the latter easier to write and understand

File: project_test/analysis_file.py (hunk 1 above)
Comment (as2388) [reply project_test/analysis_file.py]: I think we should probably use the same format everywhere.
However, the current format was chosen simply because of editor support, so feel free to raise this in tech_data_team if you want to change it.

Comment (IsaackMwenda) [reply project_test/analysis_file.py]: thanks! @as2388  I agree its good to keep it consistent across a project. PTAL.

File: project_test/analysis_file.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            for code in plan.code_scheme.codes:
+                show_matrix_keys.append(f"{plan.analysis_file_key}{code.string_value}")
+
+            AnalysisKeys.set_matrix_keys(
+                user, data, show_matrix_keys, plan.code_scheme, plan.coded_field, plan.analysis_file_key)
+
+            matrix_keys.extend(show_matrix_keys)
+                    
+        binary_keys = [plan.binary_analysis_file_key
+                        for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS
```
Comment (as2388) [inline project_test/analysis_file.py]: indentation is slightly off here

**Review by as2388 — APPROVED** [review]
